packages/celotool/src/e2e-tests/slashing_tests.ts

```diff
@@ -0,0 +1,251 @@
+
+function headerArray(web3: Web3, block: any) {
+  return [
+    block.parentHash,
+    block.sha3Uncles,
+    block.miner,
+    block.stateRoot,
+    block.transactionsRoot,
+    block.receiptsRoot,
+    block.logsBloom,
+    web3.utils.toHex(block.difficulty),
+    block.number,
+    block.gasLimit,
+    block.gasUsed,
+    block.timestamp,
+    block.extraData,
+    block.mixHash,
+    block.nonce,
+  ]
+}
+
+function headerFromBlock(web3: Web3, block: any) {
+  return rlp.encode(headerArray(web3, block))
+}
+
+describe('slashing tests', function(this: any) {
+  const gethConfig: GethTestConfig = {
+    migrate: true,
+    instances: [
+      { name: 'validator0', validating: true, syncmode: 'full', port: 30303, rpcport: 8545 },
+      { name: 'validator1', validating: true, syncmode: 'full', port: 30305, rpcport: 8547 },
+      { name: 'validator2', validating: true, syncmode: 'full', port: 30307, rpcport: 8549 },
+      { name: 'validator3', validating: true, syncmode: 'full', port: 30309, rpcport: 8551 },
+      // Validator 4 will be down in the downtime test
+      { name: 'validator4', validating: true, syncmode: 'full', port: 30311, rpcport: 8553 },
+    ],
+  }
+
+  const gethConfigDown = {
+    migrate: true,
+    instances: [
+      { name: 'validator0', validating: true, syncmode: 'full', port: 30303, rpcport: 8545 },
+      { name: 'validator1', validating: true, syncmode: 'full', port: 30305, rpcport: 8547 },
+      { name: 'validator2', validating: true, syncmode: 'full', port: 30307, rpcport: 8549 },
+      { name: 'validator3', validating: true, syncmode: 'full', port: 30309, rpcport: 8551 },
+    ],
+  }
+
+  const context: any = getContext(gethConfig)
+  const contextDown: any = getContext(gethConfigDown)
+  let web3: any
+  let kit: ContractKit
+
+  before(async function(this: any) {
+    this.timeout(0)
+    await context.hooks.before()
+  })
+
+  after(context.hooks.after)
+
+  const restart = async () => {
+    await context.hooks.restart()
+    web3 = new Web3('http://localhost:8545')
+    kit = newKitFromWeb3(web3)
+    await sleep(1)
+  }
+
+  const restartWithDowntime = async () => {
```

Comment:
we could add this feature to the `context` to avoid duplicating the config. Something like `context.hooks.partialRestart({ exclude: ['validator4']})`

Reply:
Did some simplification to config.